lib/models/BucketPolicy.js

```diff
@@ -0,0 +1,127 @@
+ *                  }
+ *              }
+ *          }
+ *      ]
+ *  }
+ */
+
+class BucketPolicy {
+    /**
+     * Create a Bucket Policy instance
+     * @param {string} json - the json policy
+     * @return {object} - BucketPolicy instance
+     */
+    constructor(json) {
+        this._json = json;
+        this._policy = {};
+    }
+
+    /**
+     * Get the bucket policy
+     * @return {object} - the bucket policy or error
+     */
+    getBucketPolicy() {
+        const policy = this._getPolicyJSON();
+        return policy;
+    }
+
+    /**
+     * Get the bucket policy array
+     * @return {object} - contains error if policy validation fails
+     */
+    _getPolicyJSON() {
+        const policy = {};
+        if (!this._json || this._json === '') {
+            policy.error = errors.MalformedPolicy.customizeDescription(
+                'request json is empty or undefined');
+            return policy;
+        }
+        const validSchema = validateResourcePolicy(this._json);
+        if (validSchema.error) {
+            policy.error = validSchema.error;
+            return policy;
+        }
+        this._getStatementArray();
+        const valAcRes = this._validateActionResource();
+        if (valAcRes.error) {
+            policy.error = valAcRes.error;
+            return policy;
+        }
+
+        return this._policy;
+    }
+
+    _getStatementArray() {
+        this._policy = JSON.parse(this._json);
+        if (!Array.isArray(this._policy.Statement)) {
+            const statement = this._policy.Statement;
+            this._policy.Statement = [statement];
+        }
+    }
+
+    /**
+     * Validate action and resource are compatible
+     * @return {error} - contains error or empty obj
+     */
+    _validateActionResource() {
+        const obj = {};
+        this._policy.Statement.forEach(s => {
+            const actions = typeof s.Action === 'string' ?
+                [s.Action] : s.Action;
+            const resources = typeof s.Resource === 'string' ?
+                [s.Resource] : s.Resource;
+            const objectAction = actions.some(a => a.includes('Object'));
+            const allAction = actions.some(a => a.includes('*'));
+            const objectResource = resources.some(r => r.includes('/'));
+            if ((objectAction && !objectResource) ||
+            (objectResource && !objectAction && !allAction)) {
+                obj.error = errors.MalformedPolicy.customizeDescription(
+                    'Action does not apply to any resource(s) in statement');
+            }
+        });
+        return obj;
+    }
+
+    /**
+     * Call resource policy schema validation function
+     * @param {object} policy - the bucket policy to validate
+     * @return {undefined}
+     */
+    static validatePolicy(policy) {
+        const validated = validateResourcePolicy(policy);
+        assert.deepStrictEqual(validated, { error: null, valid: true });
```

Comment:
We should not be throwing an assertion at run time. I would suggest returning `{ valid: true }`  and `{ error: <some error> }` for error cases.

Reply:
Actually, after looking at it again, this function follows the example of LifecycleConfig and ReplicationConfig. This validate function is only used by `BucketInfo` class, so there's no way to really return an error :/